httpx/_urlparse.py

```diff
@@ -260,8 +260,11 @@ def urlparse(url: str = "", **kwargs: typing.Optional[str]) -> ParseResult:
     # For 'path' we need to drop ? and # from the GEN_DELIMS set.
     parsed_path: str = quote(path, safe=SUB_DELIMS + ":/[]@")
     # For 'query' we need to drop '#' from the GEN_DELIMS set.
+    # We also exclude '/' because it is more robust to replace it with a percent
+    # encoding despite it not being a requirement of the spec and include '%' to
+    # prevent duplicate encoding of previously quoted items.
     parsed_query: typing.Optional[str] = (
-        None if query is None else quote(query, safe=SUB_DELIMS + ":/?[]@")
+        None if query is None else quote(query, safe=SUB_DELIMS + ":?[]@%")
```

Comment:
I was surprised that I needed to include `%` but it was needed to to prevent ` ` converted to `%20` from being changed to `%2520` in tests.

Reply:
Let me know if there's additional test coverage that would make sense for this or if we've revealed another issue.

Reply:
Hrm. Can you issue this PR without that change?

Reply:
I'll rephrase: Can you drop that bit from this pull request? It seems unrelated so it'd make sense to consider it separately.

Reply:
Yeah I can but then the _other_ change will make the tests fail i.e. this pull request will introduce a incorrect behavior. I don't understand why yet.

Reply:
> I can but then the other change will make the tests fail i.e. this pull request will introduce a incorrect behavior.

Okay. That would be worthwhile, we can review from there.

Reply:
Here are the failing tests on https://github.com/encode/httpx/pull/2723/commits/5bbacceb39d6faf885e4741633474853e6d1da43

```
_____________________________________________________________________________________________________ test_param_with_existing_escape_requires_encoding ______________________________________________________________________________________________________

    def test_param_with_existing_escape_requires_encoding():
        url = httpx.URL("http://webservice", params={"u": "http://example.com?q=foo%2Fa"})
>       assert str(url) == "http://webservice?u=http%3A%2F%2Fexample.com%3Fq%3Dfoo%252Fa"
E       AssertionError: assert 'http://webse...3Dfoo%25252Fa' == 'http://webse...q%3Dfoo%252Fa'
E         - http://webservice?u=http%3A%2F%2Fexample.com%3Fq%3Dfoo%252Fa
E         + http://webservice?u=http%253A%2F%2Fexample.com%253Fq%253Dfoo%25252Fa
E         ?                          ++                    ++    ++         ++

tests/test_urlparse.py:144: AssertionError
__________________________________________________________________________________________________________________ test_url_query_encoding ___________________________________________________________________________________________________________________

    def test_url_query_encoding():
        """
        URL query parameters should use '%20' to encoding spaces,
        and should treat '/' as a safe character. This behaviour differs
        across clients, but we're matching browser behaviour here.
    
        See https://github.com/encode/httpx/issues/2536
        and https://github.com/encode/httpx/discussions/2460
        """
        url = httpx.URL("https://www.example.com/?a=b c&d=e/f")
        assert url.raw_path == b"/?a=b%20c&d=e%2Ff"
    
        url = httpx.URL("https://www.example.com/", params={"a": "b c", "d": "e/f"})
>       assert url.raw_path == b"/?a=b%20c&d=e%2Ff"
E       AssertionError: assert b'/?a=b%2520c&d=e%2Ff' == b'/?a=b%20c&d=e%2Ff'
E         At index 7 diff: b'5' != b'0'
E         Use -v to get more diff

tests/models/test_url.py:366: AssertionError
```